### PR TITLE
Configure distribution and github actions for release (PXY-451)

### PR DESCRIPTION
**Pull request by fteychene**
## What does this PR do ?

Add maven configuration to build and publish gateway docker image
Add maven release plugin and configure it
Add distribution management to push to github packages
Add github actions workflow to build on main
Add new github action workflow to prepare the release and trigger publication of the tagged version
Add new workflow to publish artifacts and docker images from a tag or a commit

Workflow
release workflow

## Rel 

[PXY-451](https://linear.app/conduktor/issue/PXY-451/[oss]-build-and-release-action)
[PXY-414](https://linear.app/conduktor/issue/PXY-414/[oss]-release-process-for-proxy-oss)
[PXY-483](https://linear.app/conduktor/issue/PXY-483/[oss]-make-and-release-docker-image)

### CODE REVIEW

File: interceptor-framework/README.md [hunk 1]
```diff
@@ -252,7 +252,7 @@ The pom.xml in the loggerInterceptor package demonstrates one way to do this.
 
 Place the built interceptor jar file on the classpath of the gateway.
 
-Restart the gateway with the new jar file on the classpath.  In this example, the new interceptor is in a jar file that can be found in the conductor-proxy-oss-holding repository under `myNewInterceptor/target/proxy-1.0-SNAPSHOT.jar`.
+Restart the gateway with the new jar file on the classpath.  In this example, the new interceptor is in a jar file that can be found in the conductor-gateqy repository under `myNewInterceptor/target/proxy-1.0-SNAPSHOT.jar`.
```
Comment (ehumber) [inline interceptor-framework/README.md:255]: multiple typos

conduktor-gateway

File: logger-interceptor/pom.xml [hunk 2]
```diff
@@ -34,4 +35,12 @@
         </plugins>
     </build>
 
+    <distributionManagement>
+        <repository>
+            <id>github</id>
+            <name>GitHub Packages</name>
+            <url>https://maven.pkg.github.com/conduktor/conduktor-gateway</url>
```
Comment (ehumber) [inline logger-interceptor/pom.xml:42]: I'm not sure we want this.  The logger-interceptor pom.xml is supposed to be an example of what users should copy when building their own plugins.

If a user is building their own interceptor and they want it hosted on marketplace, they need to email it to Conduktor, and ask us to upload it to our mvn repo.  We shouldn't be getting them to do this themselves.

Or they are building their own interceptor and they just want that hosted in their own mvn repository.  In which case, that's not our mvn repo.

So either this needs to be a property for the URL, that has <your mvn repo here> as the property value  (though we need to be careful to ensure that we don't break our mvn package step if we do that).

Or we shouldn't include this distributionManagement step.

What do you think?

File: pom.xml [hunk 3]
```diff
@@ -100,7 +100,33 @@
                     </annotationProcessorPaths>
                 </configuration>
             </plugin>
+            <plugin>
+                <artifactId>maven-release-plugin</artifactId>
```
Comment (ehumber) [inline pom.xml:104]: Hmmm, if we publish this code, then 3rd party users can potentially run it and publish new versions of gateway?

**Review by ehumber — CHANGES_REQUESTED** [review]
See comments and questions

**Comment by ehumber** [comment]
Couple of requests

- there are a few new pre-reqs pulled in here, so we need to add their licenses to the licenses folder 

https://github.com/crazy-max/ghaction-import-gpg/blob/master/LICENSE  
https://github.com/christian-draeger/read-properties/blob/master/LICENSE
https://github.com/actions/checkout/blob/main/LICENSE
https://github.com/actions/setup-java/blob/main/LICENSE
https://github.com/actions/github-script/blob/main/LICENSE.md
https://github.com/docker/login-action/blob/master/LICENSE

 - Can you document the release a new jar + docker image process in Slite please

File: logger-interceptor/pom.xml (hunk 2 above)
Comment (fteychene) [reply logger-interceptor/pom.xml:42]: All I can think of is to only have the logger-interceptor project as a profile ofr change the dev workflow with it.
Cause it's in the same maven project but it's not the same project :shrug: 

Imho the solution is to remove it from the parent as a module and let it live as a separate project.
Or we add some profile to only add it in dev mode.
Or we specifically exclude it from the release and publishing mode. Cause if we want to release it to be able to demo it we should put it somewhere

File: pom.xml (hunk 3 above)
Comment (fteychene) [reply pom.xml:104]: If they can write on the repo yes

**Comment by fteychene** [comment]
> This is a public repository. If we inject these secrets, does that give people the ability to run github actions that can publish gateway releases (or indeed hack in a random image and push that to our docker repo), or otherwise do things we don't want them to do?

Only if they can run the actions. And they are limited to members of the repository

File: logger-interceptor/pom.xml (hunk 2 above)
Comment (fteychene) [reply logger-interceptor/pom.xml:42]: Done

**Comment by ehumber** [comment]
> > This is a public repository. If we inject these secrets, does that give people the ability to run github actions that can publish gateway releases (or indeed hack in a random image and push that to our docker repo), or otherwise do things we don't want them to do?
> 
> Only if they can run the actions. And they are limited to members of the repository

We just tested this - a third party contributor can no run gitops actions, they can not approve a review, and they can not merge.

**Review by ehumber — APPROVED** [review]

**Pull request merged**